src/pages/networks/forms/NetworkForwardForm.tsx

```diff
@@ -0,0 +1,333 @@
+import React, { FC, useEffect } from "react";
+import {
+  Button,
+  Col,
+  Form,
+  Icon,
+  Input,
+  Label,
+  Notification,
+  Row,
+  Select,
+  useNotify,
+} from "@canonical/react-components";
+import { FormikProps } from "formik/dist/types";
+import * as Yup from "yup";
+import { LxdNetworkForward } from "types/network";
+import { updateMaxHeight } from "util/updateMaxHeight";
+import useEventListener from "@use-it/event-listener";
+import { useQuery } from "@tanstack/react-query";
+import { queryKeys } from "util/queryKeys";
+import { fetchNetwork } from "api/networks";
+import { testValidIp, testValidPort } from "util/networks";
+import NotificationRow from "components/NotificationRow";
+
+export const toNetworkForward = (
+  values: NetworkForwardFormValues,
+): LxdNetworkForward => {
+  return {
+    listen_address: values.listenAddress,
+    description: values.description,
+    config: {
+      target_address: values.defaultTargetAddress,
+    },
+    ports: values.ports.map((port) => ({
+      listen_port: port.listenPort?.toString(),
+      protocol: port.protocol,
+      target_address: port.targetAddress?.toString(),
+      target_port: port.targetPort?.toString(),
+    })),
+  };
+};
+
+export const NetworkForwardSchema = Yup.object().shape({
+  listenAddress: Yup.string()
+    .test("valid-ip", "Invalid IP address", testValidIp)
+    .required("Listen address is required"),
+  ports: Yup.array().of(
+    Yup.object().shape({
+      listenPort: Yup.string()
+        .test("valid-port", "Invalid port number", testValidPort)
+        .required("Listen port required"),
+      protocol: Yup.string().required("Protocol is required"),
+      targetAddress: Yup.string()
+        .test("valid-ip", "Invalid IP address", testValidIp)
+        .required("Target address is required"),
+      targetPort: Yup.string()
+        .nullable()
+        .test("valid-port", "Invalid port number", testValidPort),
+    }),
+  ),
+});
+
+export interface NetworkForwardFormValues {
+  listenAddress: string;
+  defaultTargetAddress?: string;
+  description?: string;
+  ports: NetworkForwardPortFormValues[];
+}
+
+interface NetworkForwardPortFormValues {
+  listenPort: string;
+  protocol: "tcp" | "udp";
+  targetAddress: string;
+  targetPort?: string;
+}
+
+interface Props {
+  formik: FormikProps<NetworkForwardFormValues>;
+  isEdit?: boolean;
+  networkName: string;
+  project: string;
+}
+
+const NetworkForwardForm: FC<Props> = ({
+  formik,
+  isEdit,
+  networkName,
+  project,
+}) => {
+  const notify = useNotify();
+
+  const { data: network } = useQuery({
```

Comment:
just a nit, so feel free to disregard. Instead of making a request, we could pass network data using React router. In `NetworkForwards` component we can use <Link to={{pathname: [path-to-create-or-edit-forms], state: {network}}}>. Then in `CreateNetworkForward` and `EditNetworkForward` components we can get the state via the `location` prop. Again, feel free to skip this one

Reply:
Good point! This also reminds me we need to be consistent with queryKeys for networks, I'll refactor all the network related ones to start with `queryKeys.projects, project, queryKeys.networks`. We had some conflicts to that structure and also were introducing new conflicts in this PR. After this change, they should all align with the general structure (which always starts with `queryKeys.projects, project`. Which makes sense, because the entities live in a project as a root.

Reply:
Great, thanks for making that change! :pray: